Implement MetricAssignment form-part

jessie-ross opened the pull request:
<!-- Describe your changes in detail. -->
<img width="767" alt="Screen Shot 2020-07-26 at 11 42 59 pm" src="https://user-images.githubusercontent.com/971886/88483363-e6f54100-cf99-11ea-9d22-e4ef20ae03de.png">

Demo Vid: https://d.pr/v/4dLjMq

- Changed the form-part's title to "Goals"
- Used a percentage for Min Difference, is this what we are doing?

Part of a series on #9


## General Thoughts 

I am not too happy about the overall form of it, due to two factors:
1. The Assign A Metric being split from the Assigned Metrics can be unintuitive. 
2. Having the Metric Assignment fields in a table isn't great, it can be hard to see what field is for what and we can only provide limited info alongside each field.

I think this is good enough for an MVP but for the future I was thinking about:
- Adding a Modal step for editing and setting the Metric Assignment configuration
   This is hinted at by the original design. I don't think this is ideal as it can be annoying having to in and out of modals.
   Solves factor 1 and 2.
- Making the table fields flow vertically within a row: there would only be two cells one for the name and the other for the MetricAssignment form.
   Solves factor 1.
- Moving the Assign a Metric to the end of the AssignedMetrics Table, as the last row. 
   I tried it out for a bit but couldn't get it to look right, I am not sure what that was.
   Solves factor 2.


Also would there be any objections to changing the "24 hours" attribution window text to "1 day" for consistency with the rest?

<!-- Remember to include context: Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->


## How has this been tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
<!-- Delete any bullet points that don't apply and add more details if needed. -->

- Automated tests that cover added/changed functionality
- Manual testing with mock data (locally)

yanirs commented:
Looks good! I'll leave the full code review to @withinboredom as the assigned reviewer, but here are some thoughts.

**On correctness:**

* A metric can be assigned multiple times with different attribution windows.
* `Change expected` is independent of the the minimum difference. Experimenters always need to specify the minimum difference.
* The minimum difference is in absolute terms. It depends on the metric parameter type. See the implementation on the experiment detail page and a recent discussion on pbmo2S-oh-p2#comment-922.

**On style/content:**

* I don't think that adding the "Goals" terminology is necessary at this point, and it may be confusing. We can stick to Metrics as the section title, which is consistent with the detail page and everywhere else.
* Changing "1 day" to "24 hours" makes things a bit more ambiguous. I'd rather change "3 days" to "72 hours".
* The text feels a bit too verbose and repetitive. I played with it a bit in the browser and got to:

  ![image](https://user-images.githubusercontent.com/3952615/88496429-6e759b00-d000-11ea-9d9a-3981761fd468.png)

   Changes made:
  
    * Goals -> Metrics
    * Add -> Assign
    * Removed extra headers and repetitive text
    * Widened the metric search dropdown and changed the text to "Assign a metric to quantify the experiment's impact"

  I think this would work better, though I'm sure it can be improved. It'd be good to get feedback from @aaronyan and @withinboredom on this.

withinboredom commented on components/experiment-creation/Goals.tsx:
```diff
@@ -0,0 +1,284 @@
+                      {avaliableMetricIds
+                        .map((metricId) => normalizedMetrics[metricId])
+                        .map((metric) => (
+                          <MenuItem value={metric.metricId} key={metric.metricId}>
```
Can we add the `description` as a `title` attribute so the user can learn more about the metric by hovering over the name when selecting.

jessie-ross replied:
This is a nice idea, I'll try it out!

withinboredom commented on components/experiment-creation/Goals.tsx:
```diff
@@ -0,0 +1,284 @@
+                      id='add-metric-select'
+                      value={selectedMetricId}
+                      onChange={onSelectedMetricChange}
+                      displayEmpty
```
I didn't test what it looks like without this prop, but it looks very strange when it's empty! Can we add an easter egg here? Maybe if it's empty, we add a new unselectable metric that says something funny or even just "There's no more metrics to add."

I also wonder if it's worth adding an "Add All" button. But we should probably get some real life users constantly adding all metrics first :)

jessie-ross replied:
As Yanir mentioned above, a user should actually be able to add metric assignments with the same metric.

We will show all metrics so it will never run out 😄

withinboredom replied:
Easter egg suggestion: "Yanir Approved!" with the :yanir-approved: emoji :)

jessie-ross commented on components/experiment-creation/Goals.tsx:
```diff
@@ -0,0 +1,284 @@
+import { AttributionWindowSecondsToHuman } from '@/lib/metric-assignments'
+import MoreMenu from '../MoreMenu'
+
+const normalizedMetrics: Record<number, MetricBare> = {
```
Yeah this will be populated from the server and we should just be showing everything, see #236. We shouldn't have test data in our production server... Any chance of having another server instance with a seperate DB?

withinboredom replied:
No, probably not. There's no staging servers or anything. The best we could do is have a separate database table, and then switch to the other table somehow (in production :trollface:).

jessie-ross commented on components/experiment-creation/Goals.tsx:
```diff
@@ -0,0 +1,284 @@
+import { MetricAssignment, MetricBare, MetricParameterType, AttributionWindowSeconds } from '@/lib/schemas'
+import { TextField, Select, Switch } from 'formik-material-ui'
+import { AttributionWindowSecondsToHuman } from '@/lib/metric-assignments'
+import MoreMenu from '../MoreMenu'
```
I haven't given much thought to imports. 
This is the first project where I have used absolute imports and I have barely noticed it since `tsserver` does all the importing renaming etc.

What is the significance of using relative vs absolute?

withinboredom replied:
I personally favor less magic and like relative imports. There's some conversation on p2: p4TIVU-9sM-p2 that's related. We should be consistent though.

jessie-ross replied:
Yeah it looks like we overwhelmingly use absolutes.

(That P2 seems to be about package-relative rather than relative to this repo.)

aaronyan commented:
Exciting stuff! Looking pretty good so far @jessie-ross. Thanks for including some of your thoughts and different approaches.

Great use of the button for "change expected" by the way! I think it looks really good. I'm also a fan of the "Set as Primary" option in the ellipses.

And I think @yanirs adjustments look good! I would only adjust this part:

> Widened the metric search dropdown and changed the text to "Assign a metric to quantify the experiment's impact"

I think it looks nice to keep it simple with only the text "Assign a metric" on the dropdown form.

I think our tool can benefit from more supportive text to describe each section and do a bit more hand-holding. I think for now, we can keep it super clean, and we can figure out that extra layer of description later.

I actually kind of miss the extra words ("assigned metrics"), but I think it's just cause I like the overall look and the headers gave the card more structure. That overall look can likely be fixed by adjusting the header styling in #121 😄

yanirs commented:
> I actually like calling it "goals" instead of "metrics" and think we should change it everywhere :D it's much more intuitive for someone who doesn't need to know how everything works, where using metrics and metric assignments feels more like a leaky abstraction of the implementation.

We can discuss this separately as the consensus is to keep it out of this PR, but I don't think that avoiding the word _metric_ would make it more intuitive given the prevalent use of the word at Automattic.

> I think our tool can benefit from more supportive text to describe each section and do a bit more hand-holding. I think for now, we can keep it super clean, and we can figure out that extra layer of description later.

Yeah, in our case we're not launching to millions of users, so I think we can figure out what supportive text is needed through "manual" user testing, i.e., Zoom sessions with experimenters. It's easy to make a UI too busy with extra text that users end up ignoring anyway, and we're a bit too close to the problem now to be good judges of what people would find intuitive. I'm very much looking forward to iterating on the entire platform based on user feedback. :slightly_smiling_face:

jessie-ross commented:
## PR Update

- Changed "Goal" back to "Metrics"
- Changed `minDifference` units from `%` to either `USD` or `conversions`
- Allow more than one metricAssignment with the same metric
- Moved the assign metric below table.
   This has strange behaviour with the vertical scrolling, see #238 for a fix that generally makes it easier to deal with long and extending form-parts.
- Moved `attributionWindow` behind `name` as they together "determine" a metric assignment.
   In the future might put a schema constraint on having multiple `metricAssignments` with the same `metric`, `attributionWindow` combo.  

Demo Vid: https://d.pr/v/L3hWZa

yanirs commented:
Looks good! 

> Changed `minDifference` units from `%` to either `USD` or `conversions`

Can we please keep this consistent with the experiment detail page? It's `$` for revenue metrics and `pp` for conversion metrics (`conversions` is incorrect): https://github.com/Automattic/abacus/blob/ae9d37b77e2a13ac0d4608dab47a126cd2250672/components/MetricAssignmentsPanel.tsx#L99-L101

It'd also be good to keep the column order and headings consistent, and use the `Label` component for the primary label. I don't mind if we change the info page headings to achieve consistency, as I find the order in this PR better.

jessie-ross commented:
What does `pp` stand for?

yanirs commented:
> What does `pp` stand for?

Percentage points. Since you asked, it's probably worth spelling it out on hover.

See: https://en.wikipedia.org/wiki/Percentage_point

jessie-ross commented:
## Update

- Used `pp` and `$` as units
   I decided it didn't make sense to refactor with `MetricAssignmentsPanel.tsx` as the two cases are quite different.
- Added a tooltip for `pp`
- Reorder MetricAssignmentsPanel to match the form

yanirs commented:
> Used pp and $ as units

Looks like the ternary condition is the wrong way around: It's showing `pp` for revenue and `$` for conversions. 

Other than that, I think it looks good from a functionality/content perspective, so I'll leave the full code review to @withinboredom.

withinboredom commented on components/experiment-creation/Metrics.tsx:
```diff
@@ -0,0 +1,290 @@
+                      return (
+                        <TableRow key={index}>
+                          <TableCell>
+                            {' '}
```
It seems like this should be unnecessary.

withinboredom commented on components/experiment-creation/Metrics.tsx:
```diff
@@ -0,0 +1,290 @@
+    setSelectedMetricId(event.target.value as string)
+  }
+
+  const makeMetricAssignmentPrimary = (indexToSet: number) => {
+    metricAssignmentsFieldHelperProps.setValue(
+      metricAssignmentsField.value.map((metricAssignment, index) => ({
+        ...metricAssignment,
+        isPrimary: index === indexToSet,
+      })),
+    )
+  }
```
Would it be better to pass the `metricAssignment` instead of the index?

```suggestion
  const makeMetricAssignmentPrimary = (metricToSet) => {
    metricAssignmentsFieldHelperProps.setValue(
      metricAssignmentsField.value.map((metricAssignment) => ({
        ...metricAssignment,
        isPrimary: metricAssignment === metricToSet,
      })),
    )
  }
```

jessie-ross replied:
Ah interesting, using the object itself as the "id".

I don't see any particular advantages to doing so and it seems to me a bit less direct and less idiomatic.

And for a disadvantage: What if someone in the future immutably changes a metricAssignment before calling `makeMetricAssignmentPrimary`

More generally, in JS I would prefer to work only at the "value" level.
- Using an object like this assigns an identity to a value that is seperate from the value, the identity should be within the value so that values are comparable. Now you would have two kinds of identity - object and value, why not have one.
- It is hidden so it can't be saved, pushed to the server, or logged.

Exceptions to this might be for performance sensitive code.

withinboredom commented on components/experiment-creation/Metrics.tsx:
```diff
@@ -0,0 +1,290 @@
+                                <MenuItem value={attributionWindowSeconds} key={attributionWindowSeconds}>
+                                  {
+                                    AttributionWindowSecondsToHuman[
+                                      (attributionWindowSeconds as unknown) as AttributionWindowSeconds
```
It seems like we're abandoning some type safety here. It might be worth adding a comment as to why it's necessary and handling the case when `AttributionWindowSecondsToHuman` has an invalid index (someone may modify the html in their browser to create a custom attribution window). I don't think the API will accept a custom window, but we should handle the error gracefully in the client.

jessie-ross replied:
I've fixed the typing.

> someone may modify the html in their browser to create a custom attribution window

That user should know they are own their own.

(And we do pretty extensive validation on the client before submit):
https://github.com/Automattic/abacus/blob/416ff1ce77e56d6cbe63ae8c81e97665574b3a5e/lib/schemas.ts#L94-L108

withinboredom commented on components/experiment-creation/Metrics.tsx:
```diff
@@ -0,0 +1,290 @@
+                      <TableRow>
+                        <TableCell colSpan={5}>
+                          <Typography variant='body1' align='center'>
+                            {' '}
```
This  space also seems unnecessary.

jessie-ross replied:
Weird, I have no clue why these are here...

withinboredom commented on components/experiment-creation/Metrics.tsx:
```diff
@@ -0,0 +1,290 @@
+              <div className={classes.addMetric}>
+                <Add className={classes.addMetricAddSymbol} />
+                <FormControl className={classes.addMetricSelect}>
+                  {/* <InputLabel id='add-metric-label'>Select a Metric</InputLabel> */}
```
This can be deleted.
